client/src/controllers/KeyboardController.test.js

```diff
@@ -0,0 +1,67 @@
+import { Application } from '@hotwired/stimulus';
+import { KeyboardController } from './KeyboardController';
+
+describe('KeyboardController', () => {
+  let app;
+  const oldWindowLocation = window.location;
+
+  const setup = async (html) => {
+    document.body.innerHTML = `<main>${html}</main>`;
+
+    app = Application.start();
+    app.register('w-kbd', KeyboardController);
+
+    await Promise.resolve();
+  };
+
+  beforeAll(() => {
+    delete window.location;
+
+    window.location = Object.defineProperties(
+      {},
+      {
+        ...Object.getOwnPropertyDescriptors(oldWindowLocation),
+        assign: { configurable: true, value: jest.fn() },
+      },
+    );
+  });
+
+  afterEach(() => {
+    app?.stop();
+    jest.clearAllMocks();
+  });
+
+  describe('simulate keydown event', () => {
+    beforeEach(async () => {
+      await setup(`
+            <button
+              id="btn"
+              class="button no"
+              data-controller="w-kbd"
+              data-w-kbd-key-value="mod+j"
+            >
+              Enable
+            </button>`);
+    });
+
+    it('should call the click event', async () => {
+      const button = document.getElementById('btn');
+      button.addEventListener('click', () => {});
+      const clickMock = jest.fn();
+      HTMLButtonElement.prototype.click = clickMock;
+
+      expect(app.controllers.length).toBe(1);
+      expect(button.getAttribute('data-w-kbd-key-value')).toBe('mod+j');
+      expect(app.controllers[0].element).toBe(button);
+
+      const keyDownEvent = new KeyboardEvent('keydown', {
+        key: 'j',
+        ctrlKey: true,
+      });
+      const isDispatched = document.dispatchEvent(keyDownEvent);
+
+      expect(isDispatched).toBe(true);
+      expect(clickMock).toHaveBeenCalled();
+    });
+  });
```

Comment:
Here's a revised test, yours was very close but here's the suggestions made;

1. Update the describe/it to be more about the desired behaviour and not the fact that we are simulating/emulating anything.
2. Simplify the HTML passed into your `setup`, just use the bare minimum HTML for the example, no classes, simpler text, one line. This makes the code easier to follow.
3. Run your setup within the test itself, it makes it easier to understand what's happening and add more tests after.
4. Start with a really simple (single letter) keyboard shortcut, then add a more complicated test after.
5. Use the helper util added above `simulateKey` to both run the fake keyboard press but also check that the `preventDefault` was correctly called.
6. Assert that `buttonClickMock` has not been called at the start of your test, this way you can know that it's your fake keypress that did it.
7. Finally, we also check that the buttonClickMock was actually called on the right element, our test button `#btn` via https://jestjs.io/docs/mock-function-api#mockfnmockcontexts

```suggestion
  describe('basic keyboard shortcut usage', () => {
    it('should call the click event when the `j` key is pressed after being registered', async () => {
      expect(buttonClickMock).not.toHaveBeenCalled();

      await setup(
        `<button id="btn" data-controller="w-kbd" data-w-kbd-key-value="j">Go</button>`,
      );

      // Simulate the keydown event & check that the default was prevented
      expect(simulateKey({ keyCode: 'j' })).toHaveProperty('keypress', false);

      expect(buttonClickMock).toHaveBeenCalledTimes(1);
      expect(buttonClickMock.mock.contexts).toEqual([
        document.getElementById('btn'),
      ]);
    });
  });
```